File: 20. Valid Parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        if(s.length() < 2) return false;
        
        stack<char> stk;
        int n = s.length();
       
        for(int i = 0; i < n; ++i){
            if(s[i] == '{' || s[i] == '[' || s[i] == '('){
                stk.push(s[i]);
            }else if(s[i] == '}' || s[i] == ']' || s[i] == ')'){
                if(stk.empty()) return false;
                if(matches(stk.top(), s[i])){
                    stk.pop();
                }else return false;
                
            }
        }
        return stk.empty();
    }
    
    bool matches(char i, char j){
        if(i == '{' && j == '}') return true;
        if(i == '(' && j == ')') return true;
        if(i == '[' && j == ']') return true;
        else return false;
    }
};

File: java/20. Valid Parentheses.java
public class Solution {
    public boolean isValid(String s) {
        if(s == null || s.length() == 0 ){
            return true;
        }
        int len = s.length();
        char[] stack = new char[len];
        int top = -1;
        
        for(int i = 0; i < len; ++i){
            char ch = s.charAt(i);
            if(ch == '(' || ch == '[' || ch == '{'){
                stack[++top] = ch;
            }else{
                if(top < 0)return false;
                char prev = stack[top];
                if(prev == '(' && ch !=')'|| prev == '[' && ch != ']'
              || prev == '{' && ch != '}'){
                  return false;
              }else --top;
            }
        }
        if(top == -1){
            return true;
        }
        return false;
    }
}